private/rdf4cpp/rdf/parser/IStreamQuadIteratorSerdImpl.cpp

```diff
@@ -321,41 +324,53 @@ IStreamQuadIterator::Impl::Impl(void *stream,
         this->state_is_owned = true;
     }
 
-    serd_reader_set_strict(this->reader.get(), flags.contains(ParsingFlag::Strict));
-    serd_reader_set_error_sink(this->reader.get(), &Impl::on_error, this);
-    serd_reader_start_source_stream(this->reader.get(), read, error, stream, nullptr, 4096);
+    serd_reader_set_strict(this->reader, flags.contains(ParsingFlag::Strict));
+    serd_reader_set_error_sink(this->reader, &Impl::on_error, this);
+    serd_reader_start_source_stream(this->reader, read, error, stream, nullptr, 4096);
 }
 
 IStreamQuadIterator::Impl::~Impl() noexcept {
+    serd_reader_end_stream(this->reader);
+    serd_reader_free(this->reader);
+
     if (this->state_is_owned) {
         delete this->state;
     }
 }
 
 std::optional<nonstd::expected<IStreamQuadIterator::ok_type, IStreamQuadIterator::error_type>> IStreamQuadIterator::Impl::next() noexcept {
-    if (this->is_at_end()) [[unlikely]] {
-        return std::nullopt;
-    }
-
     while (this->quad_buffer.empty()) {
-        this->last_error = std::nullopt;
-        SerdStatus const st = serd_reader_read_chunk(this->reader.get());
-
-        if (quad_buffer.empty()) {
-            if (st != SerdStatus::SERD_SUCCESS) {
-                // was not able to read stmt, prefix or base
-
-                if (!this->last_error.has_value()) {
-                    // did not receive error either => must be eof
-                    this->end_flag = true;
-                    return std::nullopt;  // eof reached
-                }
-
-                serd_reader_skip_until_byte(this->reader.get(), '\n');
-                return nonstd::make_unexpected(*this->last_error);
-            } else if (this->last_error.has_value()) {
-                // non-fatal, artificially inserted error
-                return nonstd::make_unexpected(*this->last_error);
+        if (this->last_error.has_value()) {
+            // handle error from last time
+            if (this->last_error_requires_skip) {
+                this->last_error_requires_skip = false;
+                serd_reader_skip_until_byte(this->reader, '\n');
```

Comment:
Might jump to far for turtle. Looking for the next `.` would be better.

Reply:
Turtle is also usually written with newlines on each virtual triple, right? problem with looking for dots is that you could sometimes land inside an IRI

Reply:
yes usually. but it is not required AFAIK. Ending up inside a query sounds like a thing that could happen quite frequently. OK, then let's keep it as it is for now.